equinox/nn/_spectral_norm.py

```diff
@@ -98,17 +133,32 @@ def __init__(
         self.inference = inference
 
         weight = getattr(layer, weight_name)
-        if weight.ndim < 2:
-            raise ValueError("`weight` must be at least two-dimensional")
-        weight = jnp.reshape(weight, (weight.shape[0], -1))
-        dtype = weight.dtype
-        u_len, v_len = weight.shape
         ukey, vkey = jr.split(key)
-        u0 = jr.normal(ukey, (u_len,), dtype=dtype)
-        v0 = jr.normal(vkey, (v_len,), dtype=dtype)
-        for _ in range(15):
-            u0, v0 = _power_iteration(weight, u0, v0, eps)
+
+        if not exact:
+            if weight.ndim < 2:
+                raise ValueError("`weight` must be at least two-dimensional")
+            weight = jnp.reshape(weight, (weight.shape[0], -1))
+            dtype = weight.dtype
+            u_len, v_len = weight.shape
+            self.reverse = None
+            u0 = jr.normal(ukey, (u_len,), dtype=dtype)
+            v0 = jr.normal(vkey, (v_len,), dtype=dtype)
+            for _ in range(15):
+                u0, v0 = _power_iteration_old(weight, u0, v0, eps)
+        else:
+            assert (
+                input_shape is not None
+            ), "Must specify `input_shape` to use exact spectral norm!"
+            assert isinstance(self.layer, Callable)
```

Comment:
Assertions should be used for internal errors. User errors should be `ValueError`s. (This is just a normal Python thing.)

Also `isinstance(x, Callable)` -> `callable(x)`.